### ISSUE TITLE
React-Based DMN Editor: allowedValues not allowed for custom type

### ISSUE THREAD

**Issue by yesamer** [apache/incubator-kie-issues#1019]
It is not possible to set constraints for custom types

![Unknown-3](https://github.com/apache/incubator-kie-issues/assets/16005046/6ecfab56-bafb-4607-8e83-e5be2cae5428)

![Unknown-4](https://github.com/apache/incubator-kie-issues/assets/16005046/0aa99cd3-f7b2-4df7-b9ad-0c55f8ea4b34)

![Image](https://github.com/apache/incubator-kie-issues/assets/765578/959b6176-e4e1-4708-8a6c-62e4c483868f)

Both `allowedValues` and `typeConstraint` refers to the base type. So, a custom type seems allowed to define such constraints, that will be applied to the base type. 

This would make sense because:

1. a custom type “A” extend a base type “string” (e.g.)
2. both custom type “B” and “C” extends “A”, but they could have different constraints

**Comment by tiagobento** [comment]
IMHO, we need to discuss this a little bit more, as @ljmotta and I made assumptions that are maybe not true. Here are some questions we have:
- What happens when both types define constraints? Are they merged? What if their intersection is the empty set? E.g., a range [1..2] and a range [10..20]. It seemed weird for us that users would want to extend/especialize constraints in that way.
- What is the value added to the user if we allowed doing such hierarchical constraints? I mean, the only benefit I can see is "reusing" constraints to make them more specific in the child types. Seems like a way to introduce bad modeling, no?

Really looking for feedback on this one. Any insight is appreciated!

**Comment by tiagobento** [comment]
After a conversation between me, @yesamer, @ljmotta and @gitgabrio, there's only one question remaining: Can Struct types (I.e., itemDefinitions without a `typeRef` and with a list of `itemComponents`) have constraints directly applied to them? What about types with a `baseType` of such types? If so, how are those constraints written?

It was clear to us that all non-Struct types can be constrained, and that the constraints will be added to one another with an `and` semantic. The UI will adapt to the primitive type defined in the `typeRef` attribute at the top-level type of a given hierarchy of types.

Keep in mind that changing the `typeRef` of a type that has child types might potentially break the constraints for the child types.

**Comment by tiagobento** [comment]
Do we have any updates on this task? cc @ljmotta @gitgabrio @yesamer

**Comment by gitgabrio** [comment]
@tiagobento @ljmotta @yesamer 
There is an ongoing discussion on the OMG group.
My personal understanding (but also shared by @baldimir ) is that 
1. the spec themselves define/imply a given behavior (i.e. the usage of `?` expressions, also on inherited properties),
2. the idea is somehow enforced by the new `typeConstraints`, and the only difference with `allowedValues` is the projection inside collection elements
3. the introduction of `typeConstraint`, with its consequences, was dictated by a sort of contradiction in the specs and interpretation themselves (that somehow defined constraints as `UnaryTests` to be applied to base type, but on the other side restricted the full usage of them)
4. there was an "agreement" in the past, about interpretation of the spec, based on a sentence that has been removed from new specs

So, we are currently discussing of that. I'm new to this decisional mechanism, and I have no idea of the outcome. Regardless, the way we interpeted (and already implemented) the specs would allow much powerful behavior, and IMO it would be interesting to provide it to customers (being, anyway, backward and spec compatible)

@baldimir please correct if/where I'm wrong

**Comment by baldimir** [comment]
This is written in the spec: `An ItemDefinition element may restrict the values that are allowed from typeRef, using the allowedValues attribute.`
So if typeref references a type that is constraining something, adding allowedValues (and similarly typeConstraint) on the child type using this typeref, will further constraint the already existing constraints. 

This is further there in the spec: `When an ItemDefinition has sibling itemComponents, their values are available in the evaluation context of the UnaryTests of the allowedValues.` - This is not clear to me, I asked for clarification. 

Then there is this: `An alternative way to define an instance of ItemDefinition is as a composition of ItemDefinition elements. An instance of ItemDefinition may contain zero or more itemComponent, which are themselves ItemDefinitions. Each itemComponent in turn may be defined by either a typeRef and allowedValues or a nested itemComponent. In this way, complex types may be defined within DMN.`
In combination with the XSD, where there is a choice element, this implies to me, that there are these possible cases for using allowedValues and typeConstraint: 

1. ItemDefinition with typeRef to base type like string. - Can use allowedValues and typeConstraint. Cannot use itemComponents. Restricts the string type. 
2. ItemDefiniton with typeRef to another custom ItemDefinition. This custom ItemDefinition can also use allowedValues and typeConstraint or itemComponents freely. - Can use allowedValues and typeConstraint. Cannot use itemComponents. If the parent ItemDefinition referenced with typeRef is also using allowedValues, typeConstraint or itemComponents, all restrictions from such parent apply also in the child ItemDefinition. TLDR - If parent ItemDefinition has typeRestrictions, either from allowedValues, typeConstraint or its itemComponents, child ItemDefinition restricts the type further. See my first paragraph in this comment (`An ItemDefinition element may restrict the values that are allowed from typeRef, using the allowedValues attribute.`). 
3. ItemDefinition with itemComponents. Cannot use allowedValues or typeConstraint. Each itemComponent can have typeRestrictions and if it references a parent type with typeRef, can inherit restrictions.

**Comment by baldimir** [comment]
Here is the TCK discussion https://github.com/dmn-tck/tck/discussions/656

**Comment by ljmotta** [comment]
Regarding item 2:
>ItemDefiniton with typeRef to another custom ItemDefinition. This custom ItemDefinition can also use allowedValues and typeConstraint or itemComponents freely.

An ItemDefinition with `typeRef` cannot have `itemComponents`.

From the spec:
```
An ItemDefinition element SHALL be defined using only one of the alternative ways:
• reference to a built-in or imported typeRef, possibly restricted with allowedValues;
• composition of ItemDefinition elements
• function signature element. 
```

@baldimir @gitgabrio  Now, regarding the `An ItemDefinition element may restrict the values that are allowed from typeRef, using the allowedValues attribute.`. It isn't specified if this applies to built-in types, complex types, or both, giving that, I could enable the constraint to all cases. But as we previously discussed, we need to understand if this will work on the engine side or if this will not land for the next release.

**Comment by baldimir** [comment]
@ljmotta I wrote `"ItemDefiniton with typeRef to another custom ItemDefinition. This custom ItemDefinition can also use allowedValues and typeConstraint or itemComponents freely."` - I meant the ItemDefinition from typeRef can have itemComponents. To be more clear, here is an example: 

- ItemDefinition tPerson, containing multiple itemComponents. 
- ItemDefiniton tSlowPerson, that contains typeref="tPerson". 

Hope it is more clear. 

> It isn't specified if this applies to built-in types, complex types, or both, giving that, I could enable the constraint to all cases.

This means that it applies everywhere possible based on specification. It is defined like this. What is not specified is not in the specification.

**Comment by ljmotta** [comment]
@baldimir I understand, my mistake!

> This means that it applies everywhere possible based on specification. It is defined like this. What is not specified is not in the specification.

Yes, but the only part that is not clear, is if adding a constraint to a item definition which references a complex type will work on the engine side. 

ItemDefinition tPerson, containing multiple itemComponents (age, name).
ItemDefiniton tChildPerson, that contains typeref="tPerson, and typeConstraint ?.age in [0..18)

For what I understand this still doesn't work. Am I right?

**Comment by gitgabrio** [comment]
@ljmotta Yes, currently not enabled. I created a [PR](https://github.com/apache/incubator-kie-drools/pull/5851), but we are waiting some conclusion to the mentioned [discussion](https://github.com/dmn-tck/tck/discussions/656).
Point is (just to clarify), that IMO almost everything in the previous spec (from description, definition, and also xsd implementation) suggested the usage of inheritance and full expression for allowedValues/constraints, also on complex types; only one single sentence precluded that in some specific cases for inheritance. 
The new typeConstraint, again as I understood the [original request](https://issues.omg.org/issues/spec/DMN/1.5/fixed#issue-49990), was meant to overcome this limitation; moreover, that sentence has been removed. I hope we could come to the conclusion that the new typeConstraint would feature all the possible FEEL functionality, included inheritance and expression on complex types, keeping the old behaviour in allowedValues for backward compatibility.
Depending on that, we we'll have to adapt both frontend and eninge one way or the other.
